lib/restforce/error_code.rb

```diff
@@ -0,0 +1,408 @@
+    class MaxActiveRulesExceeded < ResponseError; end
+    class MaxApprovalStepsExceeded < ResponseError; end
+    class MaxFormulasPerRuleExceeded < ResponseError; end
+    class MaxRulesExceeded < ResponseError; end
+    class MaxRuleEntriesExceeded < ResponseError; end
+    class MaxTaskDescriptionExceeded < ResponseError; end
+    class MaxTmRulesExceeded < ResponseError; end
+    class MaxTmRuleItemsExceeded < ResponseError; end
+    class MergeFailed < ResponseError; end
+    class MissingArgument < ResponseError; end
+    class NonuniqueShippingAddress < ResponseError; end
+    class NoApplicableProcess < ResponseError; end
+    class NoAttachmentPermission < ResponseError; end
+    class NoInactiveDivisionMembers < ResponseError; end
+    class NoMassMailPermission < ResponseError; end
+    class NumberOutsideValidRange < ResponseError; end
+    class NumHistoryFieldsBySobjectExceeded < ResponseError; end
+    class OpWithInvalidUserTypeException < ResponseError; end
+    class OptedOutOfMassMail < ResponseError; end
+    class PackageLicenseRequired < ResponseError; end
+    class PlatformEventEncryptionError < ResponseError; end
+    class PlatformEventPublishingUnavailable < ResponseError; end
+    class PlatformEventPublishFailed < ResponseError; end
+    class PortalUserAlreadyExistsForContact < ResponseError; end
+    class PrivateContactOnAsset < ResponseError; end
+    class RecordInUseByWorkflow < ResponseError; end
+    class RequestRunningTooLong < ResponseError; end
+    class RequiredFieldMissing < ResponseError; end
+    class SelfReferenceFromTrigger < ResponseError; end
+    class ShareNeededForChildOwner < ResponseError; end
+    class SingleEmailLimitExceeded < ResponseError; end
+    class StandardPriceNotDefined < ResponseError; end
+    class StorageLimitExceeded < ResponseError; end
+    class StringTooLong < ResponseError; end
+    class TabsetLimitExceeded < ResponseError; end
+    class TemplateNotActive < ResponseError; end
+    class TerritoryRealignInProgress < ResponseError; end
+    class TextDataOutsideSupportedCharset < ResponseError; end
+    class TooManyApexRequests < ResponseError; end
+    class TooManyEnumValue < ResponseError; end
+    class TransferRequiresRead < ResponseError; end
+    class UnableToLockRow < ResponseError; end
+    class UnavailableRecordtypeException < ResponseError; end
+    class UndeleteFailed < ResponseError; end
+    class UnknownException < ResponseError; end
+    class UnspecifiedEmailAddress < ResponseError; end
+    class UnsupportedApexTriggerOperation < ResponseError; end
+    class UnverifiedSenderAddress < ResponseError; end
+    class WeblinkSizeLimitExceeded < ResponseError; end
+    class WeblinkUrlInvalid < ResponseError; end
+    class WrongControllerType < ResponseError; end
+
+    # Maps `errorCode`s returned from Salesforce to the exception class
+    # to be used for these errors
+    ERROR_EXCEPTION_CLASSES = {
```

Comment:
I'm worried about how you keep these two lists synchronised (e.g. defining the error class without the mapping). Unfortunately the alternative looks a bit meta-programmy right?

Throwing line lengths to the wind....

```ruby
[
  "ALL_OR_NONE_OPERATION_ROLLED_BACK" => AllOrNoneOperationRolledBack = Class.new(ResponseError)
]
```

Alternative, you could apply the registration pattern. IDEs and language servers will still recognise the classes being registered here, so you can just meta-program the error hash that you want to build:

```ruby
class ResponseError
  def self.inherited(klass)
    ERROR_CLASSES[klass.snake_case.upcase] = klass
  end
end

class StringTooLong < ResponseError; end

ERROR_CLASSES 
#=> { "STRING_TOO_LONG" => StringTooLong }
```

Reply:
What would you think of adding some specs which attempt to make sure that all of the errors are “mapped”? You could potentially use `#constants` to find all of the defined error classes and then make sure they are “mapped” in the hash. 

I like that as I would rather have meta programming in the tests than during runtime.

Reply:
If you What would you think of adding some specs which attempt to make sure that all of the errors are “mapped”? You could potentially use `#constants` to find all of the defined error classes and then make sure they are “mapped”.

I like that as I would rather have meta programming in the tests than during runtime. like that idea, I’d also be cool with number 1. I’d rather avoid mutating the constant as in number 2 it feels like we’re replacing one weird behaviour with another.

Reply:
Also using specs works!

Reply:
This is now enforced using tests.